### PR TITLE
Enrich forwardToken filter

### PR DESCRIPTION
**Pull request by irishshagua**
* Add ability to mask keys in the JSON payload for the forwardToken filter
  * Update docs in relation to forwardToken filter enhancements
  * Add tests covering new functionality

### CODE REVIEW

File: docs/reference/filters.md [hunk 1]
```diff
@@ -687,9 +687,10 @@ secureOauthTokenintrospectionAllKV("issuerURL", "", "", "k1", "v1", "k2", "v2")
 
 ## forwardToken
 
-The filter accepts a single string as an argument. The argument is the header
-name where the result of token info or token introspection is added when the
-request is passed to the backend.
+The filter accepts a variable number of string arguments but there must be at least 1 input which represents the header
+name where the result of token info or token introspection is added when the request is passed to the backend. Input
+arguments after the first, are considered to be keys within the JSON that you want to be excluded from forwarding to
+the backend service.
```
Comment (cjbooms) [inline docs/reference/filters.md]: Maybe reword to:
```
The filter takes the (string) header name as its first argument. The result of token info or token introspection is added to this header when the request is passed to the backend. Any subsequent arguments are keys within the JSON to be excluded from the header payload when forwarding to the backend service.
```

**Comment by irishshagua** [comment]
@aryszka / @szuecs 
Maybe one of you can help here? I made a relatively small change to one of the filters but it seems to have some knock-on effect to some unrelated tests. I can't reproduce these fails locally. Is it possible these could be false-negatives? Or is there some issue introduced with this code change?

**Comment by szuecs** [comment]
@irishshagua for whatever reason travis tests are very flaky and get worse now over time.

File: filters/auth/forwardtoken.go [hunk 2]
```diff
@@ -73,3 +98,20 @@ func (f *forwardTokenFilter) Request(ctx filters.FilterContext) {
...
+func stringInSlice(a string, list []string) bool {
+	for _, b := range list {
+		if b == a {
+			return true
+		}
+	}
+	return false
+}
+
+func removeKeys(data map[string]interface{}, keys []string) {
```
Comment (szuecs) [inline filters/auth/forwardtoken.go]: this can be simplified: https://play.golang.org/p/bvrzCWpiBHI

File: filters/auth/forwardtoken.go [hunk 3]
```diff
@@ -62,6 +76,17 @@ func (f *forwardTokenFilter) Request(ctx filters.FilterContext) {
 		return
 	}
 
+	if len(f.StripJsonKeys) > 0 {
+		switch typedTiMap := tiMap.(type) {
+		case map[string]interface{}:
+			removeKeys(typedTiMap, f.StripJsonKeys)
```
Comment (szuecs) [inline filters/auth/forwardtoken.go]: `fallthrough` seems to be the better option here

File: filters/auth/forwardtoken_test.go [hunk 4]
```diff
@@ -169,6 +198,7 @@ func TestForwardTokenIntrospection(t *testing.T) {
 			clientServer := httptest.NewServer(http.HandlerFunc(func(_ http.ResponseWriter, r *http.Request) {
 				if ti.oauthFilterPresent {
 					tokenInfo := r.Header.Get(ti.headerName)
+					fmt.Printf("Header Value is %s\n", tokenInfo)
```
Comment (szuecs) [inline filters/auth/forwardtoken_test.go]: this seems to be debug logging, so please drop

**Comment by szuecs** [comment]
@irishshagua small changes necessary, but thanks for the enhancement. :)

File: filters/auth/forwardtoken.go (hunk 3 above)
Comment (irishshagua) [reply filters/auth/forwardtoken.go]: compiler is complaining that `fallthrough` isn't supported for type switch.

Comment (szuecs) [reply filters/auth/forwardtoken.go]: ah ok, my parser was not good enough :)

**Comment by szuecs** [comment]
:+1:

**Comment by szuecs** [comment]
:+1:

**Comment by tkrop** [comment]
@szuecs I think it would be better to support a list of fields that should be forwarded instead of a list that should be dropped.

**Comment by tkrop** [comment]
:+1:

**Comment by szuecs** [comment]
@irishshagua do you want to change the PR to do whitelisting instead of blacklisting, as @tkrop suggested?
I think it might makes sense from security point of view, if you think it's a security feature.

**Comment by irishshagua** [comment]
Hey @szuecs and @tkrop 
My only conern about changing to a whitelist as opposed to a blacklist is that it would be a backward breaking change. The logic would be a little funky if no additional params means forward all fields and some params means only forward the listed fields. Not a major issue though, I can clearly document if this is the desired behaviour. What ye think?

**Comment by szuecs** [comment]
I think it’s fine to have:
- no Params -> send all
- params -> only send these

Because no filter is the: send nothing case, not?

**Comment by irishshagua** [comment]
> I think it’s fine to have:
> 
>     * no Params -> send all
> 
>     * params -> only send these
> 
> 
> Because no filter is the: send nothing case, not?

Ok, canged from blacklist to whitelist... cheers for the feedback.  :)

File: filters/auth/forwardtoken.go [hunk 5]
```diff
@@ -62,6 +76,17 @@ func (f *forwardTokenFilter) Request(ctx filters.FilterContext) {
...
 	}
 
+	if len(f.RetainJsonKeys) > 0 {
+		switch typedTiMap := tiMap.(type) {
+		case map[string]interface{}:
+			tiMap = retainKeys(typedTiMap, f.RetainJsonKeys)
+		case tokenIntrospectionInfo:
+			tiMap = retainKeys(typedTiMap, f.RetainJsonKeys)
+		default:
+			log.Errorf("Unexpected input type[%v] for `forwardToken` filter. Unable to apply mask", reflect.TypeOf(typedTiMap))
```
Comment (szuecs) [inline filters/auth/forwardtoken.go]: I think instead of `reflect.TypeOf` you can just use `%T` in the format string.

File: docs/reference/filters.md [hunk 6]
```diff
@@ -687,9 +687,10 @@ secureOauthTokenintrospectionAllKV("issuerURL", "", "", "k1", "v1", "k2", "v2")
 
 ## forwardToken
 
-The filter accepts a single string as an argument. The argument is the header
-name where the result of token info or token introspection is added when the
-request is passed to the backend.
+The filter takes the (string) header name as its first argument. The result of token info or token introspection is added to 
+this header when the request is passed to the backend. If there are additional arguments, these
+values are treated as a whitelisted of JSON keys to be included in the
```
Comment (aryszka) [inline docs/reference/filters.md]: "...whitelisted set of JSON keys..."?

**Comment by aryszka** [comment]
:+1:

**Comment by szuecs** [comment]
:+1:
